pkg/router/apisix.go

```diff
@@ -0,0 +1,229 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package router
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+
+	a6v2 "github.com/fluxcd/flagger/pkg/apis/apisix/v2"
+	flaggerv1 "github.com/fluxcd/flagger/pkg/apis/flagger/v1beta1"
+	clientset "github.com/fluxcd/flagger/pkg/client/clientset/versioned"
+	"go.uber.org/zap"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+// ApisixRouter is managing Apisix Route
+type ApisixRouter struct {
+	apisixClient clientset.Interface
+	logger       *zap.SugaredLogger
+	setOwnerRefs bool
+}
+
+const maxPriority = 10000
+
+// Reconcile creates or updates the Apisix Route
+func (ar *ApisixRouter) Reconcile(canary *flaggerv1.Canary) error {
+	if canary.Spec.RouteRef == nil || canary.Spec.RouteRef.Name == "" {
+		return fmt.Errorf("apisix route selector is empty")
+	}
+
+	apisixRoute, err := ar.apisixClient.ApisixV2().ApisixRoutes(canary.Namespace).Get(context.TODO(), canary.Spec.RouteRef.Name, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("APISIX route %s.%s get query error: %w",
+			canary.Spec.RouteRef.Name, canary.Namespace, err)
+	}
+
+	apisixRouteClone := apisixRoute.DeepCopy()
+	if len(apisixRouteClone.Spec.HTTP) == 0 {
+		return fmt.Errorf("APISIX route %s.%s's spec.http is empty",
+			canary.Spec.RouteRef.Name, canary.Namespace)
+	}
+
+	apexName, primaryName, canaryName := canary.GetServiceNames()
+	targetHttpRoute, targetIndex, err := ar.getTargetHttpRoute(canary, apisixRouteClone, apexName)
+	if err != nil {
+		return err
+	}
+	if len(targetHttpRoute.Backends) != 1 {
+		return fmt.Errorf("APISIX route %s.%s's http route %s only one http backend is supported",
+			canary.Spec.RouteRef.Name, canary.Namespace, targetHttpRoute.Name)
+	}
+
+	targetHttpRoute.Priority = maxPriority
+
+	primaryBackend := targetHttpRoute.Backends[0]
+	primaryBackend.ServiceName = primaryName
+	primaryWeight := 100
+	primaryBackend.Weight = &primaryWeight
+	targetHttpRoute.Backends[0] = primaryBackend
+
+	canaryWeight := 0
+	canaryBackend := a6v2.ApisixRouteHTTPBackend{
+		ServiceName:        canaryName,
+		ServicePort:        primaryBackend.ServicePort,
+		ResolveGranularity: primaryBackend.ResolveGranularity,
+		Weight:             &canaryWeight,
+		Subset:             primaryBackend.Subset,
+	}
+
+	targetHttpRoute.Backends = append(targetHttpRoute.Backends, canaryBackend)
+	apisixRouteClone.Spec.HTTP[targetIndex] = *targetHttpRoute
```

Comment:
this would lead to the canary apisix route having _all_ the http routes (including the ones that flagger doesn't care about). i'm wondering if its better to create a new apisix route object with just the http route that we need with the two different backends.
that way there'll be less configuration for users to go through. i'm not an apisix expert, so please let me know if my suggestion could cause problems.

Reply:
@tao12345666333  WDYT?   I think it's good optimization advice. Avoid stressing the apisix ingress controller and generating unnecessary data.

Reply:
SGTM.  And APISIX also supports this capability

Reply:
I have made the changes as suggested.